### PR TITLE
Fixes #58: Fixpoint computes incorrect node labeling

### LINKED ISSUE #58: Fixpoint computes incorrect node labelling
Hi,

The fixpoint module calculates surprising and (from my understanding) incorrect node labels. 

Consider the [attached](https://github.com/backtracking/ocamlgraph/files/900788/testfix.ml.txt) example. We take the divisor graph (up to 15, but that's arbitrary) and label each node with a set of integers, which initially contains the node label. The fixpoint algorithm is then instantiated to label each node with the union of the label sets of the transitively reachable nodes (i.e., we want to compute all multiples of the node in the graph), by taking set union as the join, set equality as equaity and identity as the transfer function.

_What I would expect:_
The node 10 should be labelled with {10}.
The node 2 should be labelled with {2,4,6,8,10,12,14}.

_What I get:_
The node 10 is labelled with {10}.
The node 2 is labelled with {8, 10, 12, 14}.

Note that a significant part of 2's expected label, including the original label 2, is missing!

A similar errors manifests itself when using forward propagation.

Thanks,
Johannes

### PR DESCRIPTION
**Pull request by johanneskloos**
This commit fixes the error described in issue #58 by taking the current node labeling into account when computing the meet of all incoming/outgoing edges.

### CODE REVIEW

**Comment by thizanne** [comment]
Following my message in #58, my final opinion on this bug is that your proposal is actually almost the correct way to do it.

Lengthy explanations, TLDR at the end of the message: the original code actually follows the original worklist algorithm specification, where you only join the incoming values from the predecessor nodes, not joining the previous value attached to the node being computed. The intuition behind this is that the values computed are supposed to be morally increasing anyway, so the previous value attached to this node should already be contained in the newly computed one.

However, this doesn't work for nodes having an initial value X_i which is not ⊥, that is entry nodes of the algorithm. Those are the nodes that bring some information on the equation system, so that the least fixpoint is not simply "⊥ everywhere". On the standard presentation, it causes no issue since these nodes usually have no predecessor (think of the initial state of a program, before trying to compute any statement). Then their values simply stays X_i, they are not put in the initial worklist, and since they have no predecessor, they are never put in this worklist later. 

This is not true in your examples, and does not correspond the interface of the `Fixpoint` module anyway: we have no notion of "entry nodes", and there is nothing that should prevent the moral initial nodes to have no predecessors. In this case, the algorithm should ensure that, if some node has an initial value X, then its final value is at least greater than X (I spare you the equation system details).

The correct way to fix this is to make the join (which is also called "meet" here) from the predecessors resulting values and the *original value* of the node, not its previous value. Although it could seem to make no difference (and it will indeed make no difference in the reachability or divisibility examples), it can happen that the iteration does not actually compute increasing values for the nodes (for instance in the presence of loop acceleration techniques such as widening), in which case you don't want to join with the previous value (that enforces increasing values).

As a side note, it turns out that another wrong thing in the code (whose fixing could, as a side effect, fix some occurences of the current bug) is that the successors of a computed node seem to be added to the worklist even if the value of the node did not change. This (apart from the side effect interaction with the current bug) should not lead to bugs, so let's fix this later.

TL;DR: Instead of taking the current node labelling into account, you should take the initial node labelling into account (which removes the need for the `~default` parameter iiuc).

**Comment by johanneskloos** [comment]
@thizanne I have adapted the pull request to implement the join as you recommended: We now join with the initial node labelling instead of the current node labelling. Does this implement what you envisioned?

**Comment by eponier** [comment]
I wake up after a few months, sorry. Thanks @johanneskloos for providing some implementation! @thizanne, do you think it is correct?

**Comment by thizanne** [comment]
Yes I do, thanks. Ping @yakobowski

**Comment by johanneskloos** [comment]
Ping @yakobowski: It's been three months - did you have time to look at this?

File: Makefile.in [hunk 1]
```diff
@@ -322,6 +322,9 @@ test: $(CMA) tests/test.ml
 test-bf: $(CMA) tests/test_bf.ml
 	ocaml unix.cma graphics.cma $^
 
+test-fixpoint: $(CMA) tests/test_fixpoint.ml
```
Comment (yakobowski) [inline Makefile.in]: Why do you need those depenencies?

**Comment by yakobowski** [comment]
Sorry, this completely slipped off my radar 😞 

> it can happen that the iteration does not actually compute increasing values for the nodes (for instance in the presence of loop acceleration techniques such as widening), in which case you don't want to join with the previous value (that enforces increasing values).

I agree with the diagnostic, and the fact that we don't want to merge with the previous value computed on the node. (Except that I don't see how you could have non-increasing values in presence of only widening. Did you mean widening + narrowing?)

Regarding the improvement (not adding the successors of a computed node seem to the worklist when the value of the node did not change), the code seems ok to me, if a bit contrived. We detect that `node_data = node_data'`, generate a function that always returns `None`, and `analyze_node` does not add the successor to the worklist in this case.

File: Makefile.in (hunk 1 above)
Comment (yakobowski) [reply Makefile.in]: I commented one line too high, I meant `unix.cma graphics.cma`

**Comment by yakobowski** [comment]
Ok for merging modulo the small remark on the test. Can you add yourself to the Changelog? Thanks! (And sorry for the delay.)

**Comment by johanneskloos** [comment]
Thanks!

**Comment by yakobowski** [comment]
You're welcome. Thanks to @thizanne and you!
